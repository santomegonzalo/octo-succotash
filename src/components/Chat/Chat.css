.Chat {
  display: flex;
  flex-direction: row;
  padding: 10px;
  border: 1px solid #bbb;
  min-height: 122px;
  box-sizing: border-box;
  margin: 10px 0;
  transition: box-shadow 0.5s;
}

.Chat:hover {
  box-shadow: 0px 0px 6px 0px rgba(136, 136, 136, 0.5);
}

.ChatAvatar {
  margin-right: 10px;
  border-radius: 50px;
  width: 100px;
  height: 100px;
}

.ChatData {
  display: flex;
  flex-direction: column;
  flex-grow: 1;
}

.ChatName {
  font-size: 0.875rem;
  font-weight: 600;
}

.ChatMessage {
  margin: 8px 0px;
  font-size: 0.875rem;
}

.ChatTime {
  font-size: 0.6875rem;
}
