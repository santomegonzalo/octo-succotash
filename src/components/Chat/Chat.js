import React from 'react';
import PropTypes from 'prop-types';
import { format } from 'date-fns';

import './Chat.css';

class Chat extends React.PureComponent {
	constructor(props) {
		super(props);

		this.handleMouseEnter = this.handleMouseEnter.bind(this);
		this.handleMouseLeave = this.handleMouseLeave.bind(this);

		this.state = {
			isEmailVisible: false
		};
	}

	render() {
		const { isEmailVisible } = this.state;
		const { fullName, timestamp, email, message, avatar } = this.props;

		return (
			<div
				className="Chat"
				onMouseEnter={this.handleMouseEnter}
				onMouseLeave={this.handleMouseLeave}
			>
				{avatar && (
					<img className="ChatAvatar" alt={fullName} src={avatar} />
				)}
				<div className="ChatData">
					<p className="ChatName">
						{isEmailVisible ? `${fullName} <${email}>` : fullName}
					</p>
					<p className="ChatMessage">{message}</p>
					<p className="ChatTime">
						{format(new Date(timestamp), 'DD/MM/YYYY HH:mm')}
					</p>
				</div>
			</div>
		);
	}

	handleMouseEnter() {
		this.setState({
			isEmailVisible: true
		});
	}

	handleMouseLeave() {
		this.setState({
			isEmailVisible: false
		});
	}
}

Chat.propTypes = {
	messageId: PropTypes.string,
	userId: PropTypes.string,
	fullName: PropTypes.string,
	timestamp: PropTypes.string,
	email: PropTypes.string,
	message: PropTypes.string,
	avatar: PropTypes.string
};

export default Chat;
