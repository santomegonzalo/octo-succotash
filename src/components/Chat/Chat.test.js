import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import Chat from './Chat';

describe('Chat component', () => {
	const chat = {
		avatar: 'http://dummyimage.com/100x100.jpg/dddddd/000000',
		email: 'jgomezh@cdc.gov',
		fullName: 'Jeremy Gomez',
		message:
			'Aenean auctor gravida sem. Praesent id massa id nisl venenatis lacinia.',
		messageId: '22a454b1-6659-400f-aa77-eaae34a77118',
		timestamp: '2016-04-14T16:53:55Z',
		userId: '205c9b7a-345d-4aa2-9e4c-56f05f72bbe8'
	};

	const chatNoAvatar = {
		avatar: null,
		email: 'jgomezh@cdc.gov',
		fullName: 'Jeremy Gomez',
		message:
			'Aenean auctor gravida sem. Praesent id massa id nisl venenatis lacinia.',
		messageId: '22a454b1-6659-400f-aa77-eaae34a77118',
		timestamp: '2016-04-14T16:53:55Z',
		userId: '205c9b7a-345d-4aa2-9e4c-56f05f72bbe8'
	};

	const getWrapperElement = (renderFunc, props) =>
		renderFunc(<Chat {...props} />);

	it('should render component with avatar', () => {
		expect(toJson(getWrapperElement(mount, chat))).toMatchSnapshot();
	});

	it('should render component without avatar', () => {
		expect(
			toJson(getWrapperElement(mount, chatNoAvatar))
		).toMatchSnapshot();
	});

	it('should not render image if avatar is null', () => {
		const wrapper = getWrapperElement(mount, chatNoAvatar);

		expect(wrapper.find('.ChatAvatar').exists()).toEqual(false);
	});

	it('should render email on mouse enter', () => {
		const wrapper = getWrapperElement(mount, chat);
		const setStateSpy = jest.spyOn(wrapper.instance(), 'setState');

		wrapper.simulate('mouseenter');

		expect(setStateSpy).toHaveBeenCalledTimes(1);
		expect(wrapper.find('.ChatName').text()).toContain(chat.email);
	});
});
