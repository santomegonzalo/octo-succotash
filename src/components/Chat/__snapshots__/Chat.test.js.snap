// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat component should render component with avatar 1`] = `
<Chat
  avatar="http://dummyimage.com/100x100.jpg/dddddd/000000"
  email="jgomezh@cdc.gov"
  fullName="Jeremy Gomez"
  message="Aenean auctor gravida sem. Praesent id massa id nisl venenatis lacinia."
  messageId="22a454b1-6659-400f-aa77-eaae34a77118"
  timestamp="2016-04-14T16:53:55Z"
  userId="205c9b7a-345d-4aa2-9e4c-56f05f72bbe8"
>
  <div
    className="Chat"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <img
      alt="Jeremy Gomez"
      className="ChatAvatar"
      src="http://dummyimage.com/100x100.jpg/dddddd/000000"
    />
    <div
      className="ChatData"
    >
      <p
        className="ChatName"
      >
        Jeremy Gomez
      </p>
      <p
        className="ChatMessage"
      >
        Aenean auctor gravida sem. Praesent id massa id nisl venenatis lacinia.
      </p>
      <p
        className="ChatTime"
      >
        14/04/2016 17:53
      </p>
    </div>
  </div>
</Chat>
`;

exports[`Chat component should render component without avatar 1`] = `
<Chat
  avatar={null}
  email="jgomezh@cdc.gov"
  fullName="Jeremy Gomez"
  message="Aenean auctor gravida sem. Praesent id massa id nisl venenatis lacinia."
  messageId="22a454b1-6659-400f-aa77-eaae34a77118"
  timestamp="2016-04-14T16:53:55Z"
  userId="205c9b7a-345d-4aa2-9e4c-56f05f72bbe8"
>
  <div
    className="Chat"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="ChatData"
    >
      <p
        className="ChatName"
      >
        Jeremy Gomez
      </p>
      <p
        className="ChatMessage"
      >
        Aenean auctor gravida sem. Praesent id massa id nisl venenatis lacinia.
      </p>
      <p
        className="ChatTime"
      >
        14/04/2016 17:53
      </p>
    </div>
  </div>
</Chat>
`;
